server/router/src/main/java/org/infinispan/server/router/router/impl/hotrod/handlers/SniRouteHandler.java

```diff
@@ -0,0 +1,58 @@
+package org.infinispan.server.router.router.impl.hotrod.handlers;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInitializer;
+import io.netty.handler.ssl.SniHandler;
+import io.netty.handler.ssl.SslContext;
+import io.netty.util.DomainNameMapping;
+import org.infinispan.commons.logging.LogFactory;
+import org.infinispan.server.router.RoutingTable;
+import org.infinispan.server.router.logging.RouterLogger;
+import org.infinispan.server.router.routes.hotrod.NettyHandlerRouteDestination;
+import org.infinispan.server.router.routes.Route;
+import org.infinispan.server.router.routes.SniRouteSource;
+
+import java.lang.invoke.MethodHandles;
+import java.util.List;
+import java.util.Optional;
+
+public class SniRouteHandler extends SniHandler {
+
+    private static final RouterLogger logger = LogFactory.getLog(MethodHandles.lookup().lookupClass(), RouterLogger.class);
+
+    private final RoutingTable routingTable;
+
+    public SniRouteHandler(DomainNameMapping<? extends SslContext> mapping, RoutingTable routingTable) {
+        super(mapping);
+        this.routingTable = routingTable;
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+        super.decode(ctx, in, out);
+
+        if(isHandShaked()) {
+            // At this point Netty has replaced SNIHandler (formally this) with SSLHandler in the pipeline.
+            // Now we need to add other handlers at the tail of the queue
+            logger.debug("Handshaked with hostname " + hostname());
+            Optional<Route<SniRouteSource, NettyHandlerRouteDestination>> route = routingTable.streamRoutes(SniRouteSource.class, NettyHandlerRouteDestination.class)
+                    .filter(r -> r.getRouteSource().getSniHostName().equals(this.hostname()))
+                    .findAny();
+
+            NettyHandlerRouteDestination routeDestination = route.orElseThrow(() -> logger.noRouteFound()).getRouteDesitnation();
```

Comment:
To be honest - I don't know at this point and I don't have enough `RouteSource` and `RouteDestination` implementations to check that. 

As a counterargument I can say that I iterate over routes only once per connection. And as benchmarks clearly showed - it has no impact on performance (TLS+SNI is as fast as pure TLS when doing put/get operations).

So I would prefer to keep it as is and get back to the implementation as soon as we have more route types. Then we can look at some common parts and simplify it. I don't like to do it prematurely...